File: 1124.longest-well-performing-interval.py
#
# @lc app=leetcode id=1124 lang=python3
#
# [1124] Longest Well-Performing Interval
#

# @lc code=start
class Solution:
    def longestWPI(self, hours: list[int]) -> int:
        ans, score = 0, 0
        #Store the index of the first occurence of a score
        first_occurence = {}
        for i, hour in enumerate(hours):
            if hour > 8:
                score += 1
            else:
                score -= 1
            
            if score > 0: 
                ans = i + 1
            first_occurence.setdefault(score, i)
            if score - 1 in first_occurence:
                ans = max(ans, i - first_occurence[score - 1])
        return ans
        
# @lc code=end

File: 1124.longest-well-performing-interval.java
/*
 * @lc app=leetcode id=1124 lang=java
 *
 * [1124] Longest Well-Performing Interval
 */
import java.util.HashMap;
import java.util.Map;
// @lc code=start
class Solution {
    public int longestWPI(int[] hours) {
        //! very challenging problem to solve
        int n = hours.length;
        int[] prefixSum = new int[n + 1];
        for (int i = 0; i < n; i++) {
            prefixSum[i + 1] = prefixSum[i] + (hours[i] > 8 ? 1 : -1);
        }

        int maxLength = 0;
        Map<Integer, Integer> firstOccurrence = new HashMap<>();
        for (int i = 0; i <= n; i++) {
            if (!firstOccurrence.containsKey(prefixSum[i])) {
                firstOccurrence.put(prefixSum[i], i);
            }
            if (prefixSum[i] > 0) {
                maxLength = Math.max(maxLength, i);
            } else if (firstOccurrence.containsKey(prefixSum[i] - 1)) {
                maxLength = Math.max(maxLength, i - firstOccurrence.get(prefixSum[i] - 1));
            }
        }
        return maxLength;
    }
}
// @lc code=end

